### PR TITLE
JEAN BAPTISTE ZIADE PortfolioPriceReturnTerms_PRODUCT_Prod

### PR DESCRIPTION
**Pull request by regnosys-prod-user**

### CODE REVIEW

**Comment by JBZ-Fragmos** [comment]
DESCRIPTION :

===> please see last Comment in

**Review by tomhealey-icma — CHANGES_REQUESTED** [review]
Please see comments in https://github.com/finos/common-domain-model/issues/2745.

**Review by lolabeis — DISMISSED** [review]
Proposal approved at the [CDM Derivatives Products and Business Events Working Group - March 13th, 2024](https://github.com/finos/common-domain-model/issues/2740)

File: rosetta-source/src/main/rosetta/product-template-type.rosetta [hunk 1]
```diff
@@ -341,7 +374,21 @@ type Product: <"Defines the product that is the subject of a tradable product de
...
 type Basket extends ProductBase: <"Defines a custom basket by referencing a product identifier and its consituents.">
-    basketConstituent Product (1..*) <"Identifies the constituents of the basket">
+    basketConstituent Product (0..*) <"Identifies the constituents of the basket">
+    [deprecated] // will be replaced by portfolioBasketConstituent, to be consequently be renamed basketConstituent, also the one-of condition will be removed
+    portfolioBasketConstituent BasketConstituent (0..*) <"Identifies the constituents of the basket">
+
+  condition: one-of
+
+type BasketConstituent extends Product :<"Identifies the constituents of the basket">
+ quantity QuantitySchedule (0..*)
```
Comment (lolabeis) [inline rosetta-source/src/main/rosetta/product-template-type.rosetta:384]: Attributes should also have descriptions. In particular, those descriptions should articulate the rationale for the multiple cardinality.

**Review by lolabeis — COMMENTED** [review]
Please attach release notes as a comment on your PR. They can be written directly in the .md format in the comment box in GitHub (which will render as html). You can follow the template used in most recent release notes.

**Comment by JBZ-Fragmos** [comment]
@lolabeis i've just completed missing descriptions in accordance with your previous comment, and have pushed update of the PR, please let me know if PR update has been well received and if content is now as expected ?

**Comment by PayalKhanna** [comment]
Superseded by https://github.com/finos/common-domain-model/pull/2776

**Comment by JBZ-Fragmos** [comment]
As an indication, have just pushed an update with 2 changes : 

**condition is slitghly modified as below :** 
image

**payerReceiver is added to PortfolioReturnTerms** (because depending on strategy and business case, the payerReceiver of the whole structure, that is defined at PerformancePayout level, may not necessarily be the same as inside individual legs, moreover for some strategy, it will necessarily vary inside the individual legs)
image

**Comment by JBZ-Fragmos** [comment]
**RELEASE NOTE** 

# Product Model - Portfolio Return Terms

_Background_
Some product may contain multiple individual return legs with a need to represent in same trade : **a**) an aggregated view of the payout in which underlier->Basket shall be defined, together with **b**) multiple return legs respective underlier->Security will also exist (that is a single underlier, mostly each being the same as the ones in Basket->basketConsituents). 
That is notably the case when the portfolio return is of priceReturn type, usually in connection with CFD-like business background. As an indication, other business cases may exist where the return at stake is of another kind, say varianceReturn or volatilityReturn, as part of dispersion strategy.
For the purpose of encompassing these business cases, PortfolioReturnTerms new type, which mainly _extends_ ReturnTerms, is added as a new PerformancePayout attribute, with an open cardinality (0..*) : that is where to define the multiple individual returns. As per the aggregated Basket level, if need be, this remains as of today, that is to say at PerformancePayout level.
This also comes with other minor changes as further detailed below.

_What is being released?_
* creation of new type PorfolioReturnTerms
* added new type to PerformancePayout
* added existing type  PriceSchedule to PerformancePayout (that is to replace current positioning at Return   initialValuationPrice PriceSchedule (0..1) <"Specifies the net initial valuation price(s) of the underlier at Performance Payout level. This price can be expressed either as an actual amount/currency, as a determination method, or by reference to another value specified in the swap document.">


_Review directions_
* In the Rosetta platform, select the Textual Browser and inspect the change identified above.
The changes can be reviewed in PR

**Comment by JBZ-Fragmos** [comment]
please could you manually fix the PR by also adding these 3 annotations , would avoid me to push again an other u^pdate juste for such minor clarification : 
image

**Comment by JBZ-Fragmos** [comment]
Hi @Leo ***@***.***>

Have included ReleaseNote, please feel free to update it as you wish if need be

[A screenshot of a computer  Description automatically generated]

**Comment by JBZ-Fragmos** [comment]
@lolabeis

just realized i had used erroneously QuantitySchedule instead of **NonNegativeQuantitySchedule** for the "quantity" associated with the initial/interim/finalValuationPrice, so in both places below i have fixed the issue with update of types below :

- i'm pushing right now related PR with these updates, 
- also the Release Note above has been updated accordingly

image

image

File: rosetta-source/src/main/rosetta/product-template-type.rosetta [hunk 2]
```diff
@@ -282,10 +281,12 @@ type PerformancePayout extends PayoutBase: <"Contains the necessary specificatio
 
 
 
-type PortfolioPriceReturnTerms extends ReturnTerms : <"Specifies an individual type of return of a Performance Payout, when such individual return is part of an aggregation of multiple similar returns, at Performance Payout net level.">
+type PortfolioReturnTerms extends ReturnTerms : <"Specifies an individual type of return of a Performance Payout, when such individual return is part of an aggregation of multiple similar returns, at Performance Payout level.">
 
+ payerReceiver PayerReceiver (1..1) <"Canonical representation of the payer and receiver parties applicable to each individual return leg.">
  underlier Product (1..1) <"Defines the product that is the subject of a tradable product definition, an underlying product definition, a physical exercise, a position, or other purposes.">
- quantity ResolvablePriceQuantity (0..1 ) <"Notional quantity of the underlier to which applies the PortfolioPriceReturnTerms.">
+ quantity NonNegativeQuantitySchedule (0..1 ) <"Notional quantity of the underlier to which applies to each individual return leg.">
```
Comment (lolabeis) [inline rosetta-source/src/main/rosetta/product-template-type.rosetta:288]: Remove "notional" from the description, and correct syntax typo: "to which applies" --> "which applies" (I guess)

File: rosetta-source/src/main/rosetta/product-template-type.rosetta [hunk 3]
```diff
@@ -217,15 +217,33 @@ type ReturnTerms: <"Specifies the type of return of a performance payout.">
...
+    initialValuationPrice PriceSchedule (0..1) <"Specifies the net initial valuation price(s) of the underlier at Performance Payout level. This price can be expressed either as an actual amount/currency, as a determination method, or by reference to another value specified in the swap document.">
+    [metadata address "pointsTo"=PriceQuantity->price]
+    interimValuationPrice PriceSchedule (0..1) <"Specifies the net initial valuation price(s) of the underlier at Performance Payout level. This price can be expressed either as an actual amount/currency, as a determination method, or by reference to another value specified in the swap document.">
+    [metadata address "pointsTo"=PriceQuantity->price]
+    finalValuationPrice PriceSchedule (0..1) <"Specifies the net final valuation price(s) of the underlier at Performance Payout level. This price can be expressed either as an actual amount/currency, as a determination method, or by reference to another value specified in the swap document.">
+    [metadata address "pointsTo"=PriceQuantity->price]
+
+    condition PorfolioOrStraightReturn:
```
Comment (lolabeis) [inline rosetta-source/src/main/rosetta/product-template-type.rosetta:237]: Add descriptions to conditions. It should address the "why we have this condition?" question. Maybe the last sentence you added in the `PerformancePayout` description belongs here?

File: rosetta-source/src/main/rosetta/product-template-type.rosetta [hunk 4]
```diff
@@ -217,15 +217,33 @@ type ReturnTerms: <"Specifies the type of return of a performance payout.">
...
+    required choice returnTerms, portfolioReturnTerms
+
+    condition PortofolioReturnIsMultipleReturns :
```
Comment (lolabeis) [inline rosetta-source/src/main/rosetta/product-template-type.rosetta:240]: Ditto

File: rosetta-source/src/main/rosetta/product-template-type.rosetta [hunk 5]
```diff
@@ -217,15 +217,33 @@ type ReturnTerms: <"Specifies the type of return of a performance payout.">
...
+    if portfolioReturnTerms exists
+    then portfolioReturnTerms count > 1
+
+    condition UnderlierOfPortfolioIsBasket:
```
Comment (lolabeis) [inline rosetta-source/src/main/rosetta/product-template-type.rosetta:244]: In this case, the condition description should address the question of why it only applies in certain cases (`priceReturnterms`).

Comment (JBZ-Fragmos) [reply rosetta-source/src/main/rosetta/product-template-type.rosetta:244]: do not agree to document all business cases "why" condition exists, i mean that is theoreticall exercise... and to what extent in practice : shall i re-explain all what is about various strategies etc. ? you would potentially require here long descriptions... what would be your suggestion, for instance ?

**Comment by JBZ-Fragmos** [comment]
@lolabeis i've done all as you suggested above, please check last update of PR i'm pushing it right now, hopefully you will find all your suggestions applied as you expected

**Comment by lolabeis** [comment]
@JBZ-Fragmos 👍

**Comment by hugohills-regnosys** [comment]
Replaced by https://github.com/finos/common-domain-model/pull/2776

**Pull request closed without merging**